vms/platformvm/txs/executor/camino_tx_executor.go

```diff
@@ -30,36 +36,171 @@ type CaminoProposalTxExecutor struct {
-		}
+		},
+	); err != nil {
+		return fmt.Errorf("%w: %s", errNodeSignatureMissing, err)
 	}
 	return nil
 }
 
+// verifyAddValidatorTxWithBonding carries out the validation for an AddValidatorTx
+// using bonding lock-mode instead of avax staking.
+// It returns the tx outputs that should be returned if this validator is not
+// added to the staking set.
+func verifyAddValidatorTxWithBonding(
+	backend *Backend,
+	chainState state.Chain,
+	sTx *txs.Tx,
+	tx *txs.AddValidatorTx,
+) (
+	[]*avax.TransferableOutput,
+	error,
+) {
+	// Verify the tx is well-formed
+	if err := sTx.SyntacticVerify(backend.Ctx); err != nil {
+		return nil, err
+	}
+
+	duration := tx.Validator.Duration()
+
+	switch {
+	case tx.Validator.Wght < backend.Config.MinValidatorStake:
+		// Ensure validator is staking at least the minimum amount
+		return nil, errWeightTooSmall
+
+	case tx.Validator.Wght > backend.Config.MaxValidatorStake:
+		// Ensure validator isn't staking too much
+		return nil, errWeightTooLarge
+
+	case tx.DelegationShares < backend.Config.MinDelegationFee:
+		// Ensure the validator fee is at least the minimum amount
+		return nil, errInsufficientDelegationFee
+
+	case duration < backend.Config.MinStakeDuration:
+		// Ensure staking length is not too short
+		return nil, errStakeTooShort
+
+	case duration > backend.Config.MaxStakeDuration:
+		// Ensure staking length is not too long
+		return nil, errStakeTooLong
+	}
+
+	if !backend.Bootstrapped.GetValue() {
+		return tx.Outs, nil
+	}
+
+	currentTimestamp := chainState.GetTimestamp()
+	// Ensure the proposed validator starts after the current time
+	startTime := tx.StartTime()
+	if !currentTimestamp.Before(startTime) {
+		return nil, fmt.Errorf(
+			"%w: %s >= %s",
+			errTimestampNotBeforeStartTime,
+			currentTimestamp,
+			startTime,
+		)
+	}
+
+	_, err := GetValidator(chainState, constants.PrimaryNetworkID, tx.Validator.NodeID)
+	if err == nil {
+		return nil, fmt.Errorf(
+			"attempted to issue duplicate validation for %s",
+			tx.Validator.NodeID,
+		)
+	}
+	if err != database.ErrNotFound {
+		return nil, fmt.Errorf(
+			"failed to find whether %s is a primary network validator: %w",
+			tx.Validator.NodeID,
+			err,
+		)
+	}
+
+	// Verify the flowcheck
+	if err := backend.FlowChecker.VerifyLock(
+		tx,
+		chainState,
+		tx.Ins,
+		tx.Outs,
+		sTx.Creds,
+		backend.Config.AddPrimaryNetworkValidatorFee,
+		backend.Ctx.AVAXAssetID,
+		locked.StateBonded,
+	); err != nil {
+		return nil, fmt.Errorf("%w: %s", errFlowCheckFailed, err)
+	}
+
+	// Make sure the tx doesn't start too far in the future. This is done last
+	// to allow the verifier visitor to explicitly check for this error.
+	maxStartTime := currentTimestamp.Add(MaxFutureStartTime)
+	if startTime.After(maxStartTime) {
+		return nil, errFutureStakeTime
+	}
+
+	return tx.Outs, nil
+}
+
 func (e *CaminoStandardTxExecutor) AddValidatorTx(tx *txs.AddValidatorTx) error {
-	if err := e.verifyNodeSignature(tx.NodeID()); err != nil {
```

Comment:
Why was the state check moved inside the loop?
Wasen't the plan to simply scan ins / outs for not having out LockMode and then call standard verification?